utils/preferences/tests/dtf/mod.rs

```diff
@@ -0,0 +1,84 @@
+mod data_provider;
+mod options;
+mod preferences;
+
+use data_provider::get_defaults;
+use icu_datetime::options::length;
+use icu_datetime::options::preferences::HourCycle;
+use icu_locid::{extensions_unicode_key, LanguageIdentifier, Locale};
+use icu_preferences::{preferences, Preferences};
+use options::{DayPeriod, LocaleMatcher};
+use preferences::{Calendar, NumberingSystem};
+
+preferences!(
+    DTFPreferencesBag,
```

Comment:
Nit: I'm not a fan of the `DTF` prefix. I'd prefer either `DateTimeFormatPreferencesBag` or simply `PreferencesBag` if we're okay with some module namespacing. Also, maybe we can drop the `Bag` from the end, and have:

- `DateTimeFormatPreferences` (usually `locale.into()` unless you're doing fancy resolution stuff)
- `DateTimeFormatOptions` (for settings that the developer usually specifies)

Reply:
What do you think about `DateTimeLocalePreferences`, do emphasize that this is a type-safe version of the `Locale`

Reply:
My concern is that in my mind the prefs may go beyond "locale" prefs - by locale I mean language/region related preferences with user overrides. But is "I prefer my long date pattern to be YYYY | MM _ DD" a locale preferences, or just a preference passed to date time format?

In my mental model, locale preferences are a subset of DTF preferences.

Reply:
I buy that reasoning. Let's try to establish a name that we like better than `DTFPreferencesBag` though.

Do you think the different constructors in the icu_datetime crate each need their own type, or can we have one for the whole crate? If they can all share the same one, then how about `DateTimePreferences` and define just one per crate.

If they need separate preference structs, then `DateTimeFormatPreferences`, `ZonedDateTimeFormatPreferences`, `TimeFormatPreferences`, etc. Note that this becomes a bit of a pain over FFI.

Reply:
> Do you think the different constructors in the icu_datetime crate each need their own type

Yes. I think we'll want to tailor the preferences to each constructor. Some may share (like ZonedDTF vs DTF), but some may not or even inherit.

> If they need separate preference structs, then DateTimeFormatPreferences, ZonedDateTimeFormatPreferences, TimeFormatPreferences, etc. Note that this becomes a bit of a pain over FFI.

That's what I did.

On FFI - I want to ensure we can handle FFI with minimal overhead, but I wouldn't want to compromise API experience for Rust users (or for JS, Python etc. users) for the sake of simplicity for us of building FFI.

I'd like to propose an optimal Rust design, and evaluate optimal JS API and Python API, and then see how challenging it is to build FFI that gives us all targets optimal or close to optimal. If there's one place where we should be comfortable to pay (some) of the cost it is in FFI plumbing.

Reply:
Note that it's frequent to have nested formatters. ZonedDateTimeFormat may need to invoke some subset of: DateTimeFormat, TimeZoneFormat, DateFormat, TimeFormat, and FixedDecimalFormat. I guess you're proposing that whenever we call down to another formatter, we call a `From` impl to take the relevant subset of preferences?

Note that we likely need to `.clone()` when we do this, which _could_ allocate memory since we cannot unfortunately make this `Copy` since it could contain a nonempty variants vector.

Having a single `DateTimePreferences` makes this somewhat simpler, but we do still need to `.clone().into()` at least once since `FixedDecimalFormat` is in another crate.

Reply:
> I guess you're proposing that whenever we call down to another formatter, we call a From impl to take the relevant subset of preferences?

Only if their preferences differ. I'm not sure what is the difference in preferences we want to accept for each of the DateTime formatters, but if some of them have the same preferences, we could reuse the same preferences struct. If they differ, then yes, we'd have separate structs and fallible or infallible `From` between them.

Reply:
> Only if their preferences differ

But every constructor has its own object. Therefore we need to perform a clone-into every time we have a nested constructor.

If all constructors take the same type of locale/preferences object, then we can pass them all by reference and there's less cloning.

Reply:
I don't understand how the identity of the struct type affects whether we take it by reference or ownership. Can you expand on that?

Reply:
If all of the constructors take `&Locale`, for instance, then they can pass the reference down to the nested constructors. Each constructor just queries the locale for the fields it needs.

If they take `FooPreferences`, then you need to `.into` or `.clone.into` for every child constructor.